### PR TITLE
Generate #[must_use] where needed

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
You can see the result [here](https://github.com/gtk-rs/gtk-rs-core/compare/master...GuillaumeGomez:regen-must-use?expand=1).

It is the second step of https://github.com/gtk-rs/gtk-rs-core/pull/445.

### CODE REVIEW

File: src/codegen/function.rs [hunk 1]
```diff
@@ -23,9 +25,92 @@ use std::{
...
+        // that into account. If the opposite, it means that there are two arguments in any case so
+        // we need "()" too.
+        let num_outs = std::cmp::max(num_out_args, num_out_sizes);
+        match analysis.outs.mode {
+            Mode::Normal | Mode::Combined if num_outs == 1 => {
+                let array_lengths: Vec<_> = analysis
+                    .outs
+                    .iter()
+                    .filter_map(|out| out.lib_par.array_length)
+                    .collect();
```
Comment (sdroege) [inline src/codegen/function.rs:63]: clippy doesn't like this `collect()`

File: src/codegen/function.rs [hunk 2]
```diff
@@ -23,9 +25,92 @@ use std::{
...
+    only_declaration: bool,
+    comment_prefix: &str,
+) -> Option<String> {
+    if analysis.ret.bool_return_is_error.is_some()
+        || analysis.ret.nullable_return_is_error.is_some()
+    {
+        return None;
+    }
+
+    let return_type = if !analysis.outs.is_empty() {
```
Comment (sdroege) [inline src/codegen/function.rs:41]: Don't we have exactly this code somewhere else already?

File: src/codegen/function.rs [hunk 3]
```diff
@@ -23,9 +25,92 @@ use std::{
...
+    if let Some(ref return_type) = return_type {
+        if let Some(parent_type_id) = parent_type_id {
+            if analysis.kind == library::FunctionKind::Method
+                && (!in_trait || only_declaration)
+                && parent_type_id == *return_type
+            {
+                return Some(format!("{}#[must_use]\n", comment_prefix));
+            }
+        }
+    }
```
Comment (bilelmoussaoui) [inline src/codegen/function.rs:106]: Completely unrelated to what you're working on, but this code could probably be re-used to check whether we should return Self or SomeTypeName; that should allow us to drop yet another `#[allow(some_clippy_warning)]`

File: src/codegen/function.rs [hunk 4]
```diff
@@ -23,9 +25,92 @@ use std::{
...
+            _ => None,
+        }
+    } else {
+        analysis.ret.parameter.as_ref().map(|out| out.lib_par.typ)
+    };
+    if let Some(ref return_type) = return_type {
+        if let Some(parent_type_id) = parent_type_id {
+            if analysis.kind == library::FunctionKind::Method
+                && (!in_trait || only_declaration)
+                && parent_type_id == *return_type
```
Comment (sdroege) [inline src/codegen/function.rs:101]: This should get some comments to explain what it actually does. AFAIU you add the annotation on every method that returns the type itself?

File: src/codegen/function.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/codegen/function.rs:41]: Kinda but not the same. :-/

File: src/codegen/function.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/codegen/function.rs:106]: Ah true.

File: src/codegen/function.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/codegen/function.rs:101]: Good idea.

File: src/codegen/function.rs (hunk 2 above)
Comment (sdroege) [reply src/codegen/function.rs:41]: That's surprising :) We're just building the return type as a string here or not?

Comment (GuillaumeGomez) [reply src/codegen/function.rs:41]: Here we just check if there is one returned type and if it's the same as `self`. However I could check if we could somehow merge both functions.

File: src/codegen/function.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/codegen/function.rs:63]: Ah right, need to ignore it here as well.

**Comment by GuillaumeGomez** [comment]
I reworked the code a bit to make it clearer. @bilelmoussaoui's suggestion (returning `Self` when possible) could be a nice follow-up.

File: src/codegen/function.rs (hunk 2 above)
Comment (sdroege) [reply src/codegen/function.rs]: And can we? :)

File: src/codegen/function.rs (hunk 4 above)
Comment (sdroege) [reply src/codegen/function.rs]: You didn't add a comment to the code yet

File: src/codegen/function.rs [hunk 5]
```diff
@@ -23,9 +23,30 @@ use std::{
...
+    analysis: &analysis::functions::Info,
+    in_trait: bool,
+    only_declaration: bool,
+    comment_prefix: &str,
+) -> Option<String> {
+    if let Some(parent_type_id) = parent_type_id {
+        if analysis.kind == library::FunctionKind::Method && (!in_trait || only_declaration) {
+            let outs = out_parameter_types(analysis);
+            if let [return_type] = *outs.as_slice() {
+                if parent_type_id == return_type {
```
Comment (sdroege) [inline src/codegen/function.rs:37]:
```suggestion
            if [parent_type_id] == *outs.as_slice() {
```

Maybe?

File: src/codegen/function.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/codegen/function.rs]: No. :'(

File: src/codegen/function.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/codegen/function.rs:37]: Wait that works? O.o

Comment (GuillaumeGomez) [reply src/codegen/function.rs:37]: No it doesn't work. You can check it with:

```rust
fn check(s: &[String], matcher: &String) {
    if let [matcher] = s {
        eprintln!("yolo!");
    }
}

fn main() {
    let matcher = "b".to_owned();
    check(&vec!["a".to_owned()], &matcher);
    check(&vec!["b".to_owned()], &matcher);
}
```

**Comment by GuillaumeGomez** [comment]
I added code comments. :)

File: src/codegen/function.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/codegen/function.rs:47]: Ah indeed, updating then!

**Review by sdroege — APPROVED** [review]

**Pull request merged**